receiver/otlpreceiver/otlphttp.go

```diff
@@ -16,29 +16,68 @@ package otlpreceiver
 
 import (
 	"bytes"
+	"errors"
 	"io/ioutil"
 	"net/http"
 
 	"github.com/gogo/protobuf/jsonpb"
 	"github.com/gogo/protobuf/proto"
 	"github.com/gogo/protobuf/types"
+	spb "google.golang.org/genproto/googleapis/rpc/status"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 
+	"go.opentelemetry.io/collector/internal/middleware"
 	"go.opentelemetry.io/collector/model/pdata"
 	"go.opentelemetry.io/collector/receiver/otlpreceiver/internal/logs"
 	"go.opentelemetry.io/collector/receiver/otlpreceiver/internal/metrics"
 	"go.opentelemetry.io/collector/receiver/otlpreceiver/internal/trace"
 )
 
 var jsonMarshaler = &jsonpb.Marshaler{}
+var errUnsupportedProtocol = errors.New("gRPC requires HTTP/2")
+var emptyMessage = &types.Empty{}
+
+func newGrpcResponseWriter(resp http.ResponseWriter, req *http.Request) (http.ResponseWriter, error) {
+	grpcRw := &grpcResponseWriter{innerRw: resp}
+
+	if req.ProtoMajor != 2 {
+		writeError(resp, grpcContentType, errUnsupportedProtocol, http.StatusBadRequest)
+		return nil, errUnsupportedProtocol
+	}
+
+	grpcRw.WriteHeader(int(codes.Internal))
+	compressed, err := unwrapProtoFromGrpc(req.Body)
+	if err != nil {
+		writeError(grpcRw, grpcContentType, err, http.StatusBadRequest)
+		return nil, err
+	}
+	if compressed {
+		newBody, err := middleware.NewBodyDecompressor(req.Body, req.Header.Get("grpc-encoding"))
+		if err != nil {
+			writeError(grpcRw, grpcContentType, err, http.StatusBadRequest)
+			return nil, err
+		}
+		req.Body = newBody
+	}
+	return grpcRw, nil
+}
 
 func handleTraces(
 	resp http.ResponseWriter,
 	req *http.Request,
 	contentType string,
 	tracesReceiver *trace.Receiver,
 	tracesUnmarshaler pdata.TracesUnmarshaler) {
+
+	if contentType == grpcContentType {
+		var err error
+		resp, err = newGrpcResponseWriter(resp, req)
+		if err != nil {
+			return
```

Comment:
1. Should we read and close the body even if this errors out? Just wondering if otherwise it may leak memory since the request is pending somewhere.
2. Should we return an InternalServerError to the client especially since we don't record the error in any other way here?